edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/TwinsController.cs

```diff
@@ -82,18 +97,47 @@ internal static JRaw GetRawJson(byte[] bytes)
 
         async Task<MethodResult> InvokeMethodAsync(DirectMethodRequest directMethodRequest)
         {
-            Events.ReceivedMethodCall(directMethodRequest, this.identity);
+            Events.ReceivedMethodCall(directMethodRequest);
             IEdgeHub edgeHub = await this.edgeHubGetter;
 
-            using (Metrics.TimeDirectMethod(this.identity.Id, directMethodRequest.Id))
-            {
-                DirectMethodResponse directMethodResponse = await edgeHub.InvokeMethodAsync(this.identity.Id, directMethodRequest);
-                Events.ReceivedMethodCallResponse(directMethodRequest, this.identity);
+            MethodResult methodResult;
+            string currentEdgeDeviceId = edgeHub.GetEdgeDeviceId();
 
-                MethodResult methodResult = GetMethodResult(directMethodResponse);
-
-                return methodResult;
+            if (this.TryGetActorId(out string actorDeviceId, out string actorModuleId))
+            {
+                string actorId = $"{actorDeviceId}/{actorModuleId}";
+
+                if (actorDeviceId == currentEdgeDeviceId)
+                {
+                    IHttpRequestAuthenticator authenticator = await this.authenticatorGetter;
+                    HttpAuthResult authResult = await authenticator.AuthenticateAsync(actorDeviceId, Option.Some(actorModuleId), this.HttpContext);
+
+                    if (authResult.Authenticated)
+                    {
+                        using (Metrics.TimeDirectMethod(actorDeviceId, directMethodRequest.Id))
+                        {
+                            DirectMethodResponse directMethodResponse = await edgeHub.InvokeMethodAsync(actorId, directMethodRequest);
+                            Events.ReceivedMethodCallResponse(directMethodRequest, actorId);
+
+                            methodResult = GetMethodResult(directMethodResponse);
+                        }
+                    }
+                    else
+                    {
+                        methodResult = new MethodErrorResult(HttpStatusCode.Unauthorized, authResult.ErrorMessage);
+                    }
+                }
+                else
+                {
+                    methodResult = new MethodErrorResult(HttpStatusCode.Unauthorized, "Only modules on the same device can invoke DirectMethods");
+                }
             }
+            else
+            {
+                methodResult = new MethodErrorResult(HttpStatusCode.BadRequest, $"Invalid header value for {Constants.ServiceApiIdHeaderKey}");
+            }
+
+            return methodResult;
         }
 
         async Task SendResponse(MethodResult methodResult)
```

Comment:
It seems this method can be shared across DeviceScopeController and TwinController.